core/schema/host.go

```diff
@@ -5,60 +5,102 @@ import (
 	"errors"
 	"fmt"
 
+	"github.com/containerd/containerd/labels"
 	"github.com/dagger/dagger/core"
-	"github.com/dagger/dagger/core/socket"
+	"github.com/dagger/dagger/dagql"
+	"github.com/dagger/dagger/engine/sources/blob"
+	"github.com/opencontainers/go-digest"
+	specs "github.com/opencontainers/image-spec/specs-go/v1"
 )
 
 type hostSchema struct {
-	*APIServer
-
-	host *core.Host
-	svcs *core.Services
+	srv *dagql.Server
 }
 
 var _ SchemaResolvers = &hostSchema{}
 
-func (s *hostSchema) Name() string {
-	return "host"
-}
-
-func (s *hostSchema) Schema() string {
-	return Host
-}
-
-func (s *hostSchema) Resolvers() Resolvers {
-	return Resolvers{
-		"Query": ObjectResolver{
-			"host": PassthroughResolver,
-		},
-		"Host": ObjectResolver{
-			"directory":     ToResolver(s.directory),
-			"file":          ToResolver(s.file),
-			"unixSocket":    ToResolver(s.socket),
-			"setSecretFile": ToResolver(s.setSecretFile),
-			"tunnel":        ToResolver(s.tunnel),
-			"service":       ToResolver(s.service),
-		},
-	}
+func (s *hostSchema) Install() {
+	dagql.Fields[*core.Query]{
+		dagql.Func("host", func(ctx context.Context, parent *core.Query, args struct{}) (*core.Host, error) {
+			return parent.NewHost(), nil
+		}).Doc(`Queries the host environment.`),
+		dagql.Func("blob", func(ctx context.Context, parent *core.Query, args struct {
+			Digest       string `doc:"Digest of the blob"`
+			Size         int64  `doc:"Size of the blob"`
+			MediaType    string `doc:"Media type of the blob"`
+			Uncompressed string `doc:"Digest of the uncompressed blob"`
+		}) (*core.Directory, error) {
+			blobDef, err := blob.LLB(specs.Descriptor{
+				MediaType: args.MediaType,
+				Digest:    digest.Digest(args.Digest),
+				Size:      args.Size,
+				Annotations: map[string]string{
+					labels.LabelUncompressed: args.Uncompressed, // TODO ???
+				},
+			}).Marshal(ctx)
+			if err != nil {
+				return nil, fmt.Errorf("failed to marshal blob source: %s", err)
+			}
+			return core.NewDirectory(parent, blobDef.ToPB(), "", parent.Platform, nil), nil
+		}).Doc("Retrieves a content-addressed blob."),
+	}.Install(s.srv)
+
+	dagql.Fields[*core.Host]{
+		dagql.Func("directory", s.directory).
+			Impure().
+			Doc(`Accesses a directory on the host.`).
+			ArgDoc("path", `Location of the directory to access (e.g., ".").`).
+			ArgDoc("exclude", `Exclude artifacts that match the given pattern (e.g., ["node_modules/", ".git*"]).`).
+			ArgDoc("include", `Include only artifacts that match the given pattern (e.g., ["app/", "package.*"]).`),
+
+		dagql.Func("file", s.file).
+			Impure().
+			Doc(`Accesses a file on the host.`).
+			ArgDoc("path", `Location of the file to retrieve (e.g., "README.md").`),
+
+		dagql.Func("unixSocket", s.socket).
```

Comment:
Hmm yeah worth sussing out.

On the one hand, this resolver doesn't actually have any side effects; it just returns a little wrapper object that embeds a constant string value of the path. Its sole purpose is to be passed to `Container.withUnixSocket`.

On the other hand, usage of this value implies that there were external inputs to the pipeline that aren't accounted for in a reproducability sense, i.e. it's possible to get different results when submitting the same query repeatedly.

But, what if you're mounting a socket so you can fetch a checksummed or versioned value from it? Would be a shame to bust caches all the time even when I'm using it "safely."

Maybe there's too much meaning loaded behind "pure" and really there are steps in between that are worth expressing:

* Cacheable within a single Dagger session
* 100% hermetic and reproducible
  * Even if dependent on a network, the inputs are checksummed, or at the very least versioned.
* Repeatable on my machine
* Same results on every call on any machine on my LAN
* Same results on every call on any machine on Earth
* Publishable to the Daggerverse

Let's just make it a `float64` and call it a day!

...But yeah, I'm not quite sure how to classify all this yet, open to suggestions. Maybe repeatability vs. reproducability? But boy those are hard to type.

For now I've just been focusing on the first point, which is the only concrete step we're making in this PR. Technically it implements 'save and load pipelines' but maybe the rest of the work there is to figure all of that out (and figure out how to expose it DX-wise).

What do ya think?

(Side tangent: if we figure out the taxonomy and we want to go really hardcore we could disable networking for `withExec` so it's _really_ pure and have a `withNetworkedExec` that's stamped with the same metadata as `host.unixSocket`. Bonus bikeshed for which should be the default.)

Reply:
To throw more fuel in the fire: I guess technically since the cache is shared between all modules in a session then the path of the socket is even less of a unique identifier since you could technically have different modules with unix sockets at the same path that do different things (*technically*...)

But on the other hand, as you said, the same problem exists when you have a (possibly remote/global) cache in combination with execs that have access to mutable internet/local network dependencies.

Since we currently went with the easier-to-use-but-less-hermetic model of caching execs that have access to the internet, then that does indeed feel like an argument for unix socket mounts following the same model, which would imply they stay pure unless we change everything else.

My instincts tell me that "unix socket path" is less unique than "dns name/ip address"  (just on a practical level) and so it still might make sense to mark this an impure, but I see the argument both ways.

> But, what if you're mounting a socket so you can fetch a checksummed or versioned value from it? Would be a shame to bust caches all the time even when I'm using it "safely."

Totally agree, but I would almost want to model usage like that as something similar to our existing `http` API.

Like if we wanted to support this in the most general sense, then it would make sense to have an API that says "read this file using this protocol from this unix/inet/whatever address, here's the SHA you should get for it" (or multiple APIs for all the combinations there of protocol/address type).

> What do ya think?

You're really nerd sniping me here on a Friday night 😅 I'm not gonna fully bite yet but completely unrefined thoughts:
1. We haven't had time to implement module function caching control yet, but that's going to be a place where a lot of this comes to a head. Will be interesting. Can easily imagine annotations like `// +reproducible`. Maybe it even disables internet access by default?
   * One of the ideas around implementing that was a way to turn any external state you depend on into a function input. So if you want to deploy to AWS, one of your inputs is the current state of the infra you are deploying to (so then if it hasn't changed since you last deployed, you don't re-run it). This would imply some way for users to write functions that retrieve the current state. 
   * The above somehow feels like the same-ish thing as what I mentioned above in terms of a generalization of the http API. Not exactly the same but there's some commonality there to be figured out. Maybe.
2. I'm inclined to keep the current behavior of execs having internet access by default but it should definitely taint them in some trackable way that propagates through the DAG. Supply DAG security 😎
   * Looking at how existing SBOM and provenance stuff deals with all these sorts of things would probably be enlightening
   * So then If you are a very good user who never accesses the internet in their execs, always uses hashed inputs (via one of our well controlled APIs for retrieving sources), then you definitely get a gold star ⭐️. And you'll know you did that because we can track it through the sort of generalized purity system we're discussing here.
3. If you had a global distributed remote cache that provided some degree of guarantee on retention of cached results, then in theory you could make any inputs reproducible for an arbitrary amount of time.
   * Kind of like how we currently turn local sources into blobsource and declare it valid for the duration of the session, but now generalized for longer than a single session (or engine, etc.)